contracts/contracts/buyback/Buyback.sol

```diff
@@ -52,26 +68,78 @@ contract Buyback is Governable {
      * protocol (e.g. Sushiswap)
      **/
     function swap() external onlyVault {
+        // Don't revert everything, even if the buyback fails.
+        // We want the overall transaction to continue regardless.
+        (bool success, bytes memory data) = address(this).call(
+            abi.encodeWithSignature("swapAndCheck()")
+        );
+        if (!success) {
+            emit BuybackFailed(data);
+        }
+    }
+
+    function swapAndCheck() external {
+        require(
+            msg.sender == address(this),
+            "Buyback: swapAndCheck only internal"
+        );
         if (uniswapAddr == address(0)) return;
 
         uint256 sourceAmount = ousd.balanceOf(address(this));
         if (sourceAmount < 1000 * 1e18) return;
 
-        // Uniswap redemption path
-        address[] memory path = new address[](4);
-        path[0] = address(ousd);
-        path[1] = address(usdt);
-        path[2] = IUniswapV2Router(uniswapAddr).WETH();
-        path[3] = address(ogn);
-        IUniswapV2Router(uniswapAddr).swapExactTokensForTokens(
-            sourceAmount,
-            uint256(0),
-            path,
-            address(this),
-            now
+        UniswapV3Router.ExactInputParams memory params = UniswapV3Router
+            .ExactInputParams({
+            path: abi.encodePacked(
+                ousd,
+                uint24(500), // Pool fee, ousd -> usdt
+                usdt,
+                uint24(3000), // Pool fee, usdt -> weth9
+                weth9,
+                uint24(3000), // Pool fee, weth9 -> ogn
+                ogn
+            ),
+            recipient: address(this),
+            deadline: uint256(block.timestamp + 1000),
+            amountIn: sourceAmount,
+            amountOutMinimum: uint256(0)
+        });
+
+        uint256 recieved = UniswapV3Router(uniswapAddr).exactInput(params);
+
+        // Check
+        uint256 minExpected = expectedOgnPerOUSD(sourceAmount).mul(96).div(100);
+        require(
+            recieved > minExpected,
+            "Buyback: did not receive expected OGN"
         );
     }
 
+    function expectedOgnPerOUSD(uint256 ousdAmount)
+        public
+        view
+        returns (uint256)
+    {
+        return
+            ousdAmount
+                .mul(uint256(1e26)) // ognEth is 18 decimal. ethUsd is 8 decimal.
+                .div(_price(ognEthOracle).mul(_price(ethUsdOracle)));
+    }
+
+    function _price(address _feed) internal view returns (uint256) {
+        require(_feed != address(0), "Asset not available");
+        (
+            uint80 roundID,
+            int256 _iprice,
+            uint256 startedAt,
+            uint256 timeStamp,
+            uint80 answeredInRound
+        ) = AggregatorV3Interface(_feed).latestRoundData();
+        uint256 _price = uint256(_iprice);
+        require(_price > 0, "Price must be greater than zero");
```

Comment:
Correct me if I'm wrong because I mix them up sometimes, but I don't think `_price` can ever be negative, being unsigned.  Might want to move this `require()` up one line and check against `_iprice`, as casting a negative signed to unsigned could lead to some unexpected weirdness.

Also, maybe rename one of those vars to prevent confusion between `_price` and `_iprice` when reading?

Reply:
Good points here.

Reply:
Great, should be fixed, have a look.

Reply:
Nice and clean :+1: